onnxruntime/core/optimizer/concat_slice_elimination.cc

```diff
@@ -0,0 +1,235 @@
+          steps = get_initializer_data(steps_init);
+          if (steps.size() != starts.size()) {
+            return false;
+          }
+          for (int64_t step : steps) {
+            if (step != 1) {
+              return false;
+            }
+          }
+        }
+      }
+    } else {
+      // Should be unreachable, but just to be safe in case a new op version is added.
+      return false;
+    }
+  } else {
+    return false;
+  }
+  return true;
+}
+
+/**
+Apply Concat Slice Elimination transform. This transform removes the redundant
+Concat + Slice pattern in the application of the attention bias.
+
+Before transform:
+q_bias              -- >Slice [0, q] ---> Add_q  
+      \            | 
+k_bias-- Concat --> -->Slice [q, q+k]---> Add_k
+      /            | 
+v_bias              -->Slice [q+k, :]---> Add_v
+
+After transform:
+q_bias ---> Add_q  
+      
+k_bias ---> Add_k
+
+v_bias ---> Add_v
+
+*/
+bool ConcatSliceElimination::Fuse_Subgraph(Node& concat, Graph& graph, const logging::Logger& logger) {
+  // The root could be either a graph input or a node so use node arg to compare.
+  std::vector<NodeArg*>& concat_inputs = concat.MutableInputDefs();
+  NodeArg& q_bias = *(concat_inputs[0]);
+  NodeArg& k_bias = *(concat_inputs[1]);
+  NodeArg& v_bias = *(concat_inputs[2]);
+
+  bool is_valid = graph_utils::IsInitializer(graph, q_bias.Name(), true) &&
+                  graph_utils::IsInitializer(graph, k_bias.Name(), true) &&
+                  graph_utils::IsInitializer(graph, v_bias.Name(), true);
+
+  if (!is_valid) return false;
+
+  auto get_initializer_size =
+      [&graph](const std::string& name) -> int64_t {
+    const ONNX_NAMESPACE::TensorProto* initializer = nullptr;
+    graph.GetInitializedTensor(name, initializer);
+    if (initializer != nullptr) {
+      Initializer init(*initializer, graph.ModelPath());
+        return init.size();
+    }
+    return static_cast<int64_t>(-1);
+  };
+
+  int64_t q_bias_len, k_bias_len;
+  q_bias_len = get_initializer_size(q_bias.Name());
+  k_bias_len = get_initializer_size(k_bias.Name());
+
+  std::vector<onnxruntime::Node*> node_consumers = graph.GetMutableConsumerNodes(concat.MutableOutputDefs()[0]->Name());
+  if (node_consumers.size() != 3) return false;
+
+  std::vector<onnxruntime::Node*> ordered_slice = node_consumers;
+  for (auto slice : node_consumers) {
+    std::vector<int64_t> starts, ends, axes, steps;
+    bool success = GetSliceInfo(graph, *slice, logger, starts, ends, axes, steps);
+    if (!success) return false;
+    if (starts.size() > 1) return false;
+    if (axes[0] != 0) return false;
+    if (steps[0] != 1) return false;
+    if (starts[0] == 0 && ends[0] == q_bias_len) {
+      ordered_slice[0] = slice;
+    } else if (starts[0] == q_bias_len && ends[0] == q_bias_len + k_bias_len) {
+      ordered_slice[1] = slice;
+    } else if (starts[0] == q_bias_len + k_bias_len) {
+      ordered_slice[2] = slice;
+    } else {
+      return false;
+    }
```

Comment:
The way it's written here is not very ideal. 
maybe use two counters: start and end, and check them if they matches starts[i] and ends[i] ?